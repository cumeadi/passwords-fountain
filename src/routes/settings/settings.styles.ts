import { h } from 'preact';
import { styled, setPragma } from 'goober';
import { theme, typography, media, stylelint } from '@/common/theme';
setPragma(h);

export const Wrapper = styled('section')`
    display: grid;
    justify-items: center;
    padding-top: ${theme.spacing.s12};

    ${media.gte(theme.breakpoints.m880)(stylelint.css`
        padding-top: ${theme.spacing.xl40};
    `)}
`;

export const Header = styled('header')`
    display: grid;
    justify-items: center;
    margin-bottom: ${theme.spacing.xl30};

    ${media.gte(theme.breakpoints.m880)(stylelint.css`
        margin-bottom: ${theme.spacing.xl40};
    `)}
`;

export const Heading = styled('h2')`
    text-align: center;
    color: ${theme.colors.darkBlue};
    ${typography.text24}
`;

export const FormWrapper = styled('div')`
    max-width: ${theme.breakpoints.s480};
`;

export const FormControlWrapper = styled('div')`
    padding-bottom: ${theme.spacing.xl30};
`;

export const DescriptiveText = styled('span')`
    color: ${theme.colors.darkBlue};
    font-family: ${theme.fontFamilies.regular};
`;

export const LabelWrapper = styled('div')`
    color: ${theme.colors.darkBlue};
    font-family: ${theme.fontFamilies.bold};
`;

export const NoteLabelWrapper = styled(LabelWrapper)`
    font-family: ${theme.fontFamilies.regular};
    font-style: italic;
    padding-top: ${theme.spacing.m18};
`;

export const ControlsWrapper = styled('div')`
    display: grid;
    grid-auto-flow: column;
    grid-gap: ${theme.spacing.xl30};
    padding-top: ${theme.spacing.m18};
`;
