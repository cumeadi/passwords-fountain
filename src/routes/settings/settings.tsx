import { h, VNode, Fragment } from 'preact';
import { TypedComponent } from '@/common/typings/prop-types';
import { route } from 'preact-router';
import {
    Wrapper,
    Header,
    Heading,
    FormWrapper,
    FormControlWrapper,
    DescriptiveText,
    LabelWrapper,
    NoteLabelWrapper,
    ControlsWrapper,
} from './settings.styles';
import { Text } from '@/modules/localisation/components/text';
import { useRef } from 'preact/hooks';
import { FormControl } from '@/common/components/formControl';
import { TextInput } from '@/common/components/textInput';
import { renderIfTrue } from '@/common/utils/rendering';
import { Button } from '@/common/components/button';
import { useAction, useSelector } from '@/store';
import { passwordListActions } from '@/modules/passwordList/passwordList.actions';
import { adminKey, masterKey } from '@/common/utils/formValidators';
import { useInputFormControl } from '@/common/utils/form';
import { selectIsFirstTimeOnDevice } from '@/modules/database/database.selectors';

const formValidation = {
    adminKey,
    masterKey,
} as const;

export const Settings: TypedComponent<Props> = () => {
    const isFirstTimeOnDevice = useSelector(selectIsFirstTimeOnDevice);
    const fetchPasswords = useAction(passwordListActions.fetchPasswords);
    const formRef = useRef<HTMLFormElement>(undefined as any);
    const [adminKeyInputState, adminKeyInputProps] = useInputFormControl(
        formRef,
        formValidation,
        'adminKey'
    );
    const [masterKeyInputState, masterKeyInputProps] = useInputFormControl(
        formRef,
        formValidation,
        'masterKey'
    );

    const headingText = isFirstTimeOnDevice
        ? 'settings.connectToDB'
        : 'settings.headingText';

    const handleConnectClick = async (e: Event): Promise<void> => {
        e.preventDefault();

        if (!formRef.current?.isValid) {
            return;
        }

        await fetchPasswords(
            masterKeyInputState.value,
            adminKeyInputState.value,
            true
        );
        route('/app');
    };
    const handleBackClick = (): void => {
        history.back();
    };

    const renderNoteLabel = (labelDescription: string, shouldRender: boolean) =>
        renderIfTrue(() => (
            <NoteLabelWrapper>
                <Text>settings.noteLabel</Text>{' '}
                <DescriptiveText>
                    <Text>{labelDescription}</Text>
                </DescriptiveText>
            </NoteLabelWrapper>
        ))(shouldRender);

    const renderLabel = (
        label: string,
        labelDescription: string,
        noteLabelDescription: string,
        shouldRenderNote = false
    ) => (): VNode => {
        return (
            <Fragment>
                <LabelWrapper>
                    <Text>{label}</Text> -{' '}
                    <DescriptiveText>
                        <Text>{labelDescription}</Text>
                    </DescriptiveText>
                </LabelWrapper>
                {renderNoteLabel(noteLabelDescription, shouldRenderNote)}
            </Fragment>
        );
    };
    const renderError = (error: string) => (): VNode => <Text>{error}</Text>;

    return (
        <Wrapper>
            <Header>
                <Heading>
                    <Text>{headingText}</Text>
                </Heading>
            </Header>
            <FormWrapper>
                <form ref={formRef}>
                    <FormControlWrapper>
                        <FormControl
                            hasError={Boolean(adminKeyInputProps.hasError)}
                            renderLabel={renderLabel(
                                'settings.adminKeyLabel',
                                'settings.adminKeyLabelDescription',
                                'settings.noteLabelDescriptionAdminKey'
                            )}
                            renderInput={(): VNode => (
                                <TextInput
                                    type="password"
                                    placeholder="92xIJf_ge234kalfnqql4o25ou4334201"
                                    {...adminKeyInputProps}
                                />
                            )}
                            renderError={renderError(adminKeyInputState.errors)}
                        />
                    </FormControlWrapper>
                    <FormControlWrapper>
                        <FormControl
                            hasError={Boolean(masterKeyInputProps.hasError)}
                            renderLabel={renderLabel(
                                'settings.masterKeyLabel',
                                'settings.masterKeyLabelDescription',
                                'settings.noteLabelDescription',
                                true
                            )}
                            renderInput={(): VNode => (
                                <TextInput
                                    type="password"
                                    placeholder="myMasterPassword1234"
                                    {...masterKeyInputProps}
                                />
                            )}
                            renderError={renderError(
                                masterKeyInputState.errors
                            )}
                        />
                    </FormControlWrapper>
                    <ControlsWrapper>
                        <Button onClick={handleBackClick}>
                            <Text>settings.back</Text>
                        </Button>
                        <Button
                            type="submit"
                            onClick={handleConnectClick}
                            disabled={!formRef.current?.isValid}
                        >
                            <Text>settings.connect</Text>
                        </Button>
                    </ControlsWrapper>
                </form>
            </FormWrapper>
        </Wrapper>
    );
};

export interface Props {}
