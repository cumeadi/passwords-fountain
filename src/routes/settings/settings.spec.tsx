import '@testing-library/jest-dom/extend-expect';
import { h } from 'preact';
import { render } from '@testing-library/preact';
import { Settings } from './settings';
import { StoreProvider } from '@preact-hooks/unistore';
import createStore from 'unistore';

const stateMock = {};

describe('Settings', () => {
    it('should render correctly', () => {
        const { asFragment } = render(
            <StoreProvider value={createStore(stateMock)}>
                <Settings />
            </StoreProvider>
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
