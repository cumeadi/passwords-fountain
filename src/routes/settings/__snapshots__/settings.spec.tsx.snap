// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings should render correctly 1`] = `
<DocumentFragment>
  <section
    class="go1389372715"
  >
    <header
      class="go3156958569"
    >
      <h2
        class="go1407545707"
      >
        Connect to existing database
      </h2>
    </header>
    <div
      class="go4046372879"
    >
      <form>
        <div
          class="go3855825780"
        >
          <section
            class="go3154290440"
          >
            <div
              class="go720639114"
            >
              <div
                class="go1007768492"
              >
                FaunaDB admin key - 
                <span
                  class="go3096822955"
                >
                  you can find it on your FaunaDB account.
                </span>
              </div>
            </div>
            <div
              class="go3253370986"
            >
              <input
                class="go3854179087"
                name="adminKey"
                placeholder="92xIJf_ge234kalfnqql4o25ou4334201"
                type="password"
              />
            </div>
          </section>
        </div>
        <div
          class="go3855825780"
        >
          <section
            class="go3154290440"
          >
            <div
              class="go720639114"
            >
              <div
                class="go1007768492"
              >
                Master key - 
                <span
                  class="go3096822955"
                >
                  you'll use that key every time you'll want to decrypt your passwords. Remember it!
                </span>
              </div>
              <div
                class="go1007768492 go4051580960"
              >
                Note: 
                <span
                  class="go3096822955"
                >
                  if you want to connect to database which already has encrypted passwords - use the same master key as during initial connection.
                </span>
              </div>
            </div>
            <div
              class="go3253370986"
            >
              <input
                class="go3854179087"
                name="masterKey"
                placeholder="myMasterPassword1234"
                type="password"
              />
            </div>
          </section>
        </div>
        <div
          class="go1387591026"
        >
          <button
            class="go3135434547"
            type="button"
          >
            <span
              class="go4062099715"
            >
              Back
            </span>
          </button>
          <button
            class="go3135434547"
            disabled=""
            type="submit"
          >
            <span
              class="go4062099715"
            >
              Connect
            </span>
          </button>
        </div>
      </form>
    </div>
  </section>
</DocumentFragment>
`;
